src/Compilers/CSharp/Test/Symbol/Symbols/CovariantReturnTests.cs

```diff
@@ -0,0 +1,235 @@
+                );
+            CreateCompilation(source, parseOptions: TestOptions.WithCovariantReturns).VerifyDiagnostics(
+                );
+        }
+
+        [Fact]
+        public void CovariantReturns_02()
+        {
+            var source = @"
+class Base
+{
+    public virtual T M<T, U>() where T : class where U : class, T => null;
+}
+class Derived : Base
+{
+    public override U M<T, U>() => null;
+}
+";
+            CreateCompilation(source).VerifyDiagnostics(
+                // (8,23): error CS8652: The feature 'covariant returns' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public override U M<T, U>() => null;
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "M").WithArguments("covariant returns").WithLocation(8, 23)
+                );
+            CreateCompilation(source, parseOptions: TestOptions.WithCovariantReturns).VerifyDiagnostics(
+                );
+        }
+
+        [Fact]
+        public void CovariantReturns_03()
+        {
+            var source = @"
+class Base<T> where T : class
+{
+    public virtual T M() => null;
+}
+class Derived<T, U> : Base<T> where T : class where U : class, T
+{
+    public override U M() => null;
+}
+";
+            CreateCompilation(source).VerifyDiagnostics(
+                // (8,23): error CS8652: The feature 'covariant returns' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public override U M() => null;
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "M").WithArguments("covariant returns").WithLocation(8, 23)
+                );
+            CreateCompilation(source, parseOptions: TestOptions.WithCovariantReturns).VerifyDiagnostics(
+                );
+        }
+
+        [Fact]
+        public void CovariantReturns_04()
+        {
+            var source = @"
+class N { }
+class Base
+{
+    public virtual N M() => null;
+}
+class Derived<T> : Base where T : N
+{
+    public override T M() => null;
+}
+";
+            CreateCompilation(source).VerifyDiagnostics(
+                // (9,23): error CS8652: The feature 'covariant returns' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public override T M() => null;
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "M").WithArguments("covariant returns").WithLocation(9, 23)
+                );
+            CreateCompilation(source, parseOptions: TestOptions.WithCovariantReturns).VerifyDiagnostics(
+                );
+        }
+
+        [Fact]
+        public void CovariantReturns_05()
+        {
+            var source = @"
+class Base
+{
+    public virtual object M => null;
+}
+class Derived : Base
+{
+    public override string M => null;
```

Comment:
What is expected for `public new string M => null;` scenario?

Reply:
(7): This is for `override` only, thereby excluding interfaces.

Will add comments for suggested tests. Many of these are existing tests elsewhere in our test suite, and since the diagnostics have not changed (i.e. not suggesting upgrade to language version that supports covariant returns) that is evidence that the behavior has not changed.

---
In reply to: [413973483](https://github.com/dotnet/roslyn/pull/43576#discussion_r413973483) [](ancestors = 413973483)

Reply:
We should also have a test with at least three levels:

```C#
class A
{
    public virtual object M() => null!;
}
class B : A
{
   public override string M() => null!;
}
class C : B
{
   public override object M() => null!; //error
}
class D : C
{
   public override string M() => null!; // probably fine
}
```